py-polars/polars/_cpu_check.py

```diff
@@ -1,13 +1,238 @@
+                self.win = ctypes.CDLL("kernel32.dll")
+                opc = _WINDOWS_64_OPC
+            else:
+                # Here ctypes.windll.kernel32 is needed to get the
+                # right DLL. Otherwise it will fail when running
+                # 32 bit Python on 64 bit Windows.
+                self.win = ctypes.windll.kernel32  # type: ignore[attr-defined]
+                opc = _CDECL_32_OPC
+        else:
+            opc = _POSIX_64_OPC if _IS_64BIT else _CDECL_32_OPC
+
+        size = len(opc)
+        code = (ctypes.c_ubyte * size)(*opc)
+
+        if _IS_WINDOWS:
+            self.win.VirtualAlloc.restype = c_void_p
+            self.win.VirtualAlloc.argtypes = [
+                ctypes.c_void_p,
+                ctypes.c_size_t,
+                ctypes.c_ulong,
+                ctypes.c_ulong,
+            ]
+            self.addr = self.win.VirtualAlloc(
+                None, size, _MEM_COMMIT | _MEM_RESERVE, _PAGE_EXECUTE_READWRITE
+            )
+            if not self.addr:
+                raise MemoryError("could not allocate memory for CPUID check")
+            ctypes.memmove(self.addr, code, size)
+        else:
+            import mmap  # Only import if necessary.
+
+            self.mmap = mmap.mmap(
+                -1,
+                size,
+                mmap.MAP_PRIVATE | mmap.MAP_ANONYMOUS,
+                mmap.PROT_READ | mmap.PROT_WRITE | mmap.PROT_EXEC,
+            )
+            self.addr = ctypes.addressof(ctypes.c_void_p.from_buffer(self.mmap))
+            self.mmap.write(code)
+
+        func_type = CFUNCTYPE(None, POINTER(CPUID_struct), c_uint32, c_uint32)
+        self.func_ptr = func_type(self.addr)
+
+    def __call__(self, eax: int, ecx: int = 0) -> CPUID_struct:
+        struct = CPUID_struct()
+        self.func_ptr(struct, eax, ecx)
+        return struct
+
+    def __del__(self) -> None:
+        if _IS_WINDOWS:
+            self.win.VirtualFree.restype = c_long
+            self.win.VirtualFree.argtypes = [c_void_p, c_size_t, c_ulong]
+            self.win.VirtualFree(self.addr, 0, _MEM_RELEASE)
+
+
+def read_cpu_flags() -> dict[str, bool]:
+    # Right now we only enable extra feature flags for x86.
+    if not _IS_X86:
+        return {}
+
+    # CPU flags from https://en.wikipedia.org/wiki/CPUID
+    cpuid = CPUID()
+    cpuid1 = cpuid(1, 0)
+    cpuid7 = cpuid(7, 0)
+    cpuid81h = cpuid(0x80000001, 0)
+
+    return {
+        "sse3": bool(cpuid1.ecx & (1 << 0)),
+        "ssse3": bool(cpuid1.ecx & (1 << 9)),
+        "fma": bool(cpuid1.ecx & (1 << 12)),
+        "sse4.1": bool(cpuid1.ecx & (1 << 19)),
+        "sse4.2": bool(cpuid1.ecx & (1 << 20)),
+        "popcnt": bool(cpuid1.ecx & (1 << 23)),
+        "avx": bool(cpuid1.ecx & (1 << 28)),
+        "bmi1": bool(cpuid7.ebx & (1 << 3)),
+        "bmi2": bool(cpuid7.ebx & (1 << 8)),
+        "avx2": bool(cpuid7.ebx & (1 << 5)),
+        "lzcnt": bool(cpuid81h.ecx & (1 << 5)),
+    }
+
+
+def check_cpu_flags() -> None:
```

Comment:
The variable is an empty string for development builds, so we should skip the check in that case.

Reply:
Yeah got caught by Python's `split` implementation not being empty for an empty string.